mozilla/bedrock
# Move Product Stories to Contentful (fixes #12438)

## One-line summary

Pull content for /stories and individual story pages from Contentful. 

## Significant changes and points to review

Front-end:
- new Contentful templates for landing and story pages
- new Contentful style folder contains some new styles (particularly single story grid layout) but is mostly reorganization and porting over existing styles

Back-end:
- new Views logic to query Contentful Product Stories
- new Contentful API logic to format data from Contentful into context dictionary for use in jinja templates

## Issue / Bugzilla link

https://github.com/mozilla/bedrock/issues/12438

## Checklist

If relevant:

There will be tests! This is a two-part review PR. First part is checking the Contentful Content Model and Bedrock API are structured the way we want. Once those expectations are confirmed, I'll go ahead with testing they function as expected.

- [x] Tests added (switch and block renderer in api)
- [x] Feature flags added to www-config: https://github.com/mozmeao/www-config/pull/513/files

## Testing

Demo server URL: (or None)

To test this work:

In Contentful sandbox, there are several new content modals

Story & Contributor (specific to Product Stories)
https://app.contentful.com/spaces/w5er3c7zdgmd/environments/sandbox/content_types?order.fieldId=name&order.direction=ascending&searchTerm=product

EMBED Blockquote and Image (available in rich text editors)
https://app.contentful.com/spaces/w5er3c7zdgmd/environments/sandbox/content_types?order.fieldId=name&order.direction=ascending&searchTerm=embed

- [ ] the models have the right fields and validation
- [ ] the models adhere to the overall structure of our Contentful integration

 Local testing:
- check you have `CONTENTFUL_ENVIRONMENT=sandbox`
- use `./manage.py update_contentful --force` to update data from the sandbox

landing
http://localhost:8000/en-US/stories/
https://www.mozilla.org/en-US/stories/
story
http://localhost:8000/en-US/stories/raising-technology-eq/
http://localhost:8000/en-US/stories/joy-of-color/
https://www.mozilla.org/en-US/stories/raising-technology-eq/
https://www.mozilla.org/en-US/stories/joy-of-color/

- [ ] templates are semantically correct
- [ ] styles are comparable to production
- [ ] content is displayed as expected (images, quotes, cards) ⚠️ note: this is copy-pasted dummy content, not final
- [ ] with `SWITCH_CONTENTFUL_PRODUCT_STORY=off`, the hardcoded landing page and stories in production are used without error

I would highly recommend playing with the content in Contentful sandbox (adding a story, changing the content or theme colors). There are two ways to verify that your sandbox changes are working:
1. (for landing page) re-run `./manage.py update_contentful --force` 
2. (for story pages) use the contentful preview URLS, i.e. (where the final path segment is the Entry ID from Contentful)
`localhost:8000/en-US/contentful-preview/4QR8FPH2Zx0j1n71vWYn5p/`

## Suggested next steps:
- Create real articles in Contentful
- Migrate Contentful sandbox changes
- Test Previews and Scheduled Content
- Make any code updates needed and have final review of Bedrock changes (with master env content)
- Merge Bedrock changes
- Turn switch for Contentful PBJ ON
